### PR TITLE
[CMAKE][MKSHELLLINK] Fix ReadMe icon on LiveCD

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

JIRA issue: [CORE-15156](https://jira.reactos.org/browse/CORE-15156)

## Proposed changes

- Enhance `sdk/tools/mkshelllink/mkshelllink.c` for negative icon index, text file icons, and percents.
- Fix `Read Me` shortcut icon.

## TODO

- [x] Do tests.

## Screenshots

LiveCD (AFTER):
after

### CODE REVIEW

File: boot/bootdata/CMakeLists.txt [hunk 1]
```diff
@@ -69,11 +69,20 @@ macro(add_livecd_shortcut name app dest)
...
+                TARGET livecd_links
+                DESTINATION ${dest}
+                FOR livecd)
+endmacro(add_livecd_shortcut_ex)
 
 ## NOTE: What would be nice is to create this list using /media/inf/shortcuts.inf
 ## and taking their default english translation!
 add_cd_file(FILE "${CMAKE_CURRENT_SOURCE_DIR}/livecd_start.cmd" DESTINATION "Profiles/Default User/My Documents" FOR livecd)
-add_livecd_shortcut("Read Me" readme.txt "Profiles/All Users/Desktop")
+add_livecd_shortcut_ex("Read Me" readme.txt "Profiles/All Users/Desktop" "!SystemRoot!/system32/shell32.dll" -152)
```
Comment (binarymaster) [inline boot/bootdata/CMakeLists.txt:85]: I believe this is a problem of mkshelllink, it should assign an icon automatically based on the filetype in such cases.

Comment (katahiromz) [reply boot/bootdata/CMakeLists.txt]: Fixed.

File: sdk/tools/mkshelllink/mkshelllink.c [hunk 2]
```diff
@@ -195,6 +210,13 @@ int main(int argc, const char *argv[])
         return 0;
     }
 
+    // The icon of text files
+    if (strstr(pszIcon, ".txt"))
```
Comment (SergeGautherie) [inline sdk/tools/mkshelllink/mkshelllink.c:214]: This should be good enough.
Possible improvements, I guess:
compare to end of string only,
be case-insensitive.

Comment (katahiromz) [reply sdk/tools/mkshelllink/mkshelllink.c]: Improved.

**Review by whindsaks — CHANGES_REQUESTED** [review]
The shortcut is currently broken because the icon path is incorrectly set to "Readme.txt". The command line argument has been interpreted as the icon.

----

You should not have to specify the icon at all, the .lnk icon handler will look it up. If you want to modify MKSHELLLINK, it should be changed to support a target path like `shell:Windows\Readme.txt`. This would also remove the need for the silly batch hack.

screenshot

**Comment by katahiromz** [comment]
> target path like shell:Windows

Can you analyse target path on Linux?

**Comment by katahiromz** [comment]
Improper fix. Closed.

**Comment by katahiromz** [comment]
I assign @whindsaks you to https://jira.reactos.org/browse/CORE-19692 .
Good luck.

**Comment by katahiromz** [comment]
Reopened.

`CShellLink::GetIconLocation` cannot get the icon location of LiveCD shortcut.
The solution you are claiming is impossible.

**Comment by katahiromz** [comment]
@whindsaks

> You should not have to specify the icon at all, the .lnk icon handler will look it up.

`mkshelllink` creates a shell link in standard C way. That doesn't use icon handler.
`CShellLink::GetIconLocation` cannot get the icon location of Livecd shortcut without icon info.

**Comment by whindsaks** [comment]
> `CShellLink::GetIconLocation` cannot get the icon location of Livecd shortcut without icon info.

Yes it can, it just needs a valid pidl.

mkshelllink needs to either store a environment path `%windir%\readme.exe` or the special folder CSIDL_WINDOWS pidl offset in the extra block.

**Comment by whindsaks** [comment]
It should be possible to re-open and complete this PR now.

**Pull request closed without merging**